packages/connect-node/src/http2-session-manager.ts

```diff
@@ -714,15 +714,12 @@ function ready(
       options.pingIntervalMs = options.pingIntervalMs * 2;
       receivedGoAwayEnhanceYourCalmTooManyPings = true;
     }
+
+    // Node.js closes a connection only when it receives a GOAWAY NO_ERROR, we need to destroy session ourself
+    // see https://github.com/nodejs/node/blob/v10.3.0/lib/internal/http2/core.js#L470 https://github.com/nodejs/node/blob/main/lib/internal/http2/core.js#L682
     if (errorCode === http2.constants.NGHTTP2_NO_ERROR) {
       receivedGoAwayNoError = true;
-      const nodeMajor = parseInt(process.versions.node.split(".")[0], 10);
-      // Node.js v16 closes a connection on its own when it receives a GOAWAY
-      // frame and there are no open streams (emitting a "close" event and
-      // destroying the session), but more recent versions do not.
-      // Calling close() ourselves is ineffective here - it appears that the
-      // method is already being called, see https://github.com/nodejs/node/blob/198affc63973805ce5102d246f6b7822be57f5fc/lib/internal/http2/core.js#L681
-      if (streamCount == 0 && nodeMajor >= 18) {
+      if (streamCount == 0) {
```

Comment:
As we found out in https://github.com/bufbuild/connect-es/issues/683#issuecomment-1648209277, Node.js _does_ close connections when it receives a GOAWAY with any error, but it does so after request streams have closed.

The special casing for v18 and above is because behavior differs for GOAWAY with NO_ERROR for connections without streams:

```bash
$ node -v
v16.20.0
$ node repro-goaway-close.mjs
http2.connect
conn on connect
conn on goaway, code: 0
conn on close
^C
$ node -v
v18.17.0
$ node repro-goaway-close.mjs
http2.connect
conn on connect
conn on goaway, code: 0
^C
```


<details>
<summary>repro-683.mjs</summary>

```js
import * as http2 from "http2";

// start a server that sends a GOAWAY with NO_ERROR to every session
const PORT = 8088;
const server = http2.createServer()
  .on("session", (sess) => {
    setTimeout(() => {
      sess.goaway(http2.constants.NGHTTP2_NO_ERROR)
    }, 100);
  })
  .on("request", () => {
    //
  });
await new Promise(resolve => {
  server.listen(PORT, () => resolve());
});


// open a connection
const conn = await new Promise((resolve, reject) => {
  console.log("http2.connect");
  const conn = http2.connect(`http://localhost:${PORT}`);
  conn.on("error", reject);
  conn.on("connect", () => {
    console.log("conn on connect");
    resolve(conn);
  });
  conn.on("goaway", errorCode => {
    console.log("conn on goaway, code:", errorCode);
  });
  conn.on("close", () => {
    console.log("conn on close")
  });
  conn.on("error", e => {
    console.log("conn on error", String(e));
  });
});
```
</details>

Note that v16 emits "close", v18 does not. By calling `conn.destroy()` ourselves, we force emit "error" and "close" events. You would think that calling `conn.close()` would help, but it does not. That is what the comment was referring to. We should not change this part.

Reply:
Reverted in https://github.com/bufbuild/connect-es/pull/712/commits/4df734e6c49c33105f2ded8971cab13dbe43fbae